rpc/serve.go

```diff
@@ -0,0 +1,69 @@
+package rpc
+
+import (
+	"context"
+	"errors"
+	"net"
+
+	"capnproto.org/go/capnp/v3"
+)
+
+// Serve serves a Cap'n Proto RPC to incoming connections.
+//
+// Serve will take ownership of bootstrapClient and release it after the listener closes.
+//
+// Serve exits with the listener error if the listener is closed by the owner.
+func Serve(lis net.Listener, bootstrapClient capnp.Client) error {
+	if !bootstrapClient.IsValid() {
+		err := errors.New("BootstrapClient is not valid")
+		return err
+	}
+	// Accept incoming connections
+	defer bootstrapClient.Release()
+	for {
+		conn, err := lis.Accept()
+		if err != nil {
+			// Since we took ownership of the bootstrap client, release it after we're done.
+			if !bootstrapClient.IsValid() {
+				err = errors.New("the bootstrap client was already released")
+			}
+			return err
+		}
+
+		// the RPC connection takes ownership of the bootstrap interface and will release it when the connection
+		// exits, so use AddRef to avoid releasing the provided bootstrap client capability.
+		opts := Options{
+			BootstrapClient: bootstrapClient.AddRef(),
+		}
+
+		// For each new incoming connection, create a new RPC transport connection that will serve incoming RPC requests
+		// rpc.Options will contain the bootstrap capability
+		go func() {
+			transport := NewStreamTransport(conn)
+			conn := NewConn(transport, &opts)
+			<-conn.Done()
```

Comment:
Come to think of it, you don't even need to spawn a goroutine here; instead, just skip the call to .Done() and you can do it in the main loop; it doesn't block.